### PR TITLE
Use pyproject.toml for project metadata, add citation info

### PR DESCRIPTION
**Pull request by mwtoews**
This PR moves project metadata setup.py to pyproject.toml as described by [PEP 621](https://peps.python.org/pep-0621/) and implemented by [setuptools >= 61](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html).

While most of the conversion is 1:1, here are some differences:

- Add "3.11" Python version classifier
- Remove "Operating System" classifiers, since they are all common
- Expand keywords with "bmi" and "xmi"
- Remove "platform" which is not actively used by setuptools/pip/pypi
- Add Publication URL for PyPI
- Add "dev" extras, which is an alias for "lint,test", i.e. `pip install -e .[dev]` should be good for devs
- Remove hard-link to develop branches for `xmipy` and `modflow-devtools`, as these interfere with environments with these requirements met (i.e. it will force them to be overwritten); may need to find a better way around this.

Other changes:

- Add citation info in README.md and add a CITATION.cff file for both software and preferred citations
- Add black configuration to pyproject.toml
- Update guide-to-publish.md to modern conventions
- Remove flake8 configuration from setup.cfg, since [.flake8](https://github.com/MODFLOW-USGS/modflowapi/blob/develop/.flake8) already exists (even if a bit outdated)
- Remove etc/requirements.pip.txt since this is repeated from pyproject.toml, and appears to be out-of-date
- Use Python version `3.x` (currently 3.11) for `std_setup` CI job
- Add Python 3.11 to relevant CI tests
- Remove (probably?) unnecessary Windows-specific install steps
- Add package name in setup.py for GitHub project dependencies (which is currently [empty](https://github.com/MODFLOW-USGS/modflowapi/network/dependents))

### CODE REVIEW

**Comment by jdhughes-usgs** [comment]
@mwtoews could you update the PR so that it passes the lint step?

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -171,21 +161,14 @@ jobs:
...
-          pip install .
-          pip install ".[test]"
-
-      - name: Install Python dependencies Windows
-        if: runner.os == 'Windows'
-        run: |
-          python.exe -m pip install --upgrade pip
-          python.exe -m pip install .
-          python.exe -m pip install ".[test]"
+          pip install git+https://git@github.com/Deltares/xmipy@develop
```
Comment (mwtoews) [inline .github/workflows/ci.yml:169]: is this still necessary? []v1.3.0](https://github.com/Deltares/xmipy/releases/tag/v1.3.0) was released last week.

Comment (jdhughes-usgs) [reply .github/workflows/ci.yml:169]: I could see that it is a good idea for modflowapi CI to use `xmipy@develop`. @mjr-deltares and @Hofer-Julian what do you think?

Comment (Hofer-Julian) [reply .github/workflows/ci.yml:169]: I agree that it makes sense to use develop in the CI

**Pull request merged**